Add initial capacities for brick vecs

Kmschr opened the pull request:
Slightly improve performance by avoiding re-allocations during building of brick vectors.

8 bytes was chosen as the most appropriate for the bit brick vector based on following consideration for minimum bits needed to represent a brick in save version 10. The minimum required is 7 bytes per brick in the case where all "palettes" are 1 or 2 values at most and where positions are small and bricks are not procedural. 8 bytes is a better middle ground for realistic use cases and roundness.

```
minimum bit brick
asset index - 1
brick size - 1 (not procedural)
position - 24 (0, 0, 0)
orientation - 5
collision bits - 4
visibility - 1
material index - 1
physical index - 1
intensity - 4
color - 2 (small palette)
owner - 8

total - 52 bits
```

possible pitfall is if a save file was maliciously constructed to have a max value brick count it would over allocate when reading, but this should not be a concern for brick building files.

Kmschr committed:
add initial capacities for brick vecs

voximity commented on src/read.rs:
```diff
@@ -270,7 +270,7 @@ impl<R: Read> SaveReader<R> {
         let physical_material_count = cmp::max(header2.physical_materials.len(), 2);
 
-        let mut bricks = vec![];
+        let mut bricks = Vec::with_capacity(header1.brick_count as usize);
```
To address your pitfall: it might be worth adding a `const MAX_BRICK_COUNT_CAPACITY: usize` to this module that dictates the maximum reasonable amount of bricks to allocate when reading. Maybe 2,000,000 or so. Then, you could `with_capacity` the maximum between the save-specified brick count and that constant.

voximity replied:
Also, if you're concerned about too much being allocated for values *less* than this, consider [`shrink_to_fit`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.shrink_to_fit) after the vector is populated.

voximity commented on src/write.rs:
```diff
@@ -143,7 +143,7 @@ impl<W: Write> SaveWriter<W> {
         // write bricks and components
         {
-            let mut vec = vec![];
+            let mut vec = Vec::with_capacity(self.data.bricks.len() * 8);
```
I like your reasoning for 8 bytes -- could you also throw this into a `const` in this module and add a doc comment explaining your reasoning? You can link to this PR if you want. I think it would be helpful for future readers

Kmschr committed:
Create constant for bytes per brick

Kmschr committed:
limit initial capacity to 10 million bricks, and shrink over allocated vec

Kmschr committed:
use correct name in vec call

voximity committed:
fix: unconditionally call shrink_to_fit regardless of specified brick count

This is basically free and probably a good habit anyway. I would imagine brs-js doesn't always write a brick count into the header so this would only sometimes call `shrink_to_fit`

voximity commented:
I made one small change to unconditionally call `shrink_to_fit` no matter what after reading in some bricks -- it makes sense to have a small brick buffer, since I would imagine most use cases don't add bricks (causing a reallocation).

voximity commented:
Otherwise LGTM! Thanks!

Kmschr commented:
> I made one small change to unconditionally call `shrink_to_fit` no matter what after reading in some bricks -- it makes sense to have a small brick buffer, since I would imagine most use cases don't add bricks (causing a reallocation).

`shrink_to_fit` actually performs the same check as I was making previously, there is no change to the underlying buffer when capacity is equal to length, which will be the case for all reads of <= 10 million bricks where brick count header does not lie. So I agree with this change but not for the reason you gave

```rust
pub fn shrink_to_fit(&mut self) {
    // The capacity is never less than the length, and there's nothing to do when
    // they are equal, so we can avoid the panic case in `RawVec::shrink_to_fit`
    // by only calling it with a greater capacity.
    if self.capacity() > self.len {
        self.buf.shrink_to_fit(self.len);
    }
}
```

voximity commented:
> > I made one small change to unconditionally call `shrink_to_fit` no matter what after reading in some bricks -- it makes sense to have a small brick buffer, since I would imagine most use cases don't add bricks (causing a reallocation).
> 
> `shrink_to_fit` actually performs the same check as I was making previously, there is no change to the underlying buffer when capacity is equal to length, which will be the case for all reads of <= 10 million bricks where brick count header does not lie. So I agree with this change but not for the reason you gave
> 
> ```rust
> pub fn shrink_to_fit(&mut self) {
>     // The capacity is never less than the length, and there's nothing to do when
>     // they are equal, so we can avoid the panic case in `RawVec::shrink_to_fit`
>     // by only calling it with a greater capacity.
>     if self.capacity() > self.len {
>         self.buf.shrink_to_fit(self.len);
>     }
> }
> ```

Right, that was a rather useless comment; I meant with respect to how it previously functioned _without_ a capacity
